library/common/http/dispatcher.cc

```diff
@@ -0,0 +1,105 @@
+#include "library/common/http/dispatcher.h"
+
+#include "library/common/buffer/utility.h"
+#include "library/common/http/header_utility.h"
+
+namespace Envoy {
+namespace Http {
+
+Dispatcher::DirectStreamCallbacks::DirectStreamCallbacks(envoy_stream_t stream,
+                                                         envoy_observer observer,
+                                                         Dispatcher& http_dispatcher)
+    : stream_(stream), observer_(observer), http_dispatcher_(http_dispatcher) {}
+
+void Dispatcher::DirectStreamCallbacks::onHeaders(HeaderMapPtr&& headers, bool end_stream) {
+  ENVOY_LOG(debug, "response headers for stream (end_stream={}):\n{}", end_stream, *headers);
+  if (end_stream) {
+    http_dispatcher_.removeStream(stream_);
+  }
+  observer_.h(stream_, Utility::transformHeaders(std::move(headers)), end_stream);
+}
+
+void Dispatcher::DirectStreamCallbacks::onData(Buffer::Instance& data, bool end_stream) {
+  ENVOY_LOG(debug, "response data for stream (length={} end_stream={})", data.length(), end_stream);
+  if (end_stream) {
+    http_dispatcher_.removeStream(stream_);
+  }
+  observer_.d(stream_, Envoy::Buffer::Utility::transformData(data), end_stream);
+}
+
+void Dispatcher::DirectStreamCallbacks::onTrailers(HeaderMapPtr&& trailers) {
+  ENVOY_LOG(debug, "response trailers for stream:\n{}", *trailers);
+  http_dispatcher_.removeStream(stream_);
+  observer_.t(stream_, Utility::transformHeaders(std::move(trailers)));
+}
+
+void Dispatcher::DirectStreamCallbacks::onReset() {
+  http_dispatcher_.removeStream(stream_);
+  observer_.e(stream_, {ENVOY_STREAM_RESET, {0, nullptr}});
+}
+
+Dispatcher::DirectStream::DirectStream(AsyncClient::Stream* underlying_stream,
+                                       DirectStreamCallbacksPtr callbacks)
+    : underlying_stream_(underlying_stream), callbacks_(std::move(callbacks)) {}
+
+Dispatcher::Dispatcher(Event::Dispatcher& event_dispatcher,
+                       Upstream::ClusterManager& cluster_manager)
+    : event_dispatcher_(event_dispatcher), cluster_manager_(cluster_manager) {}
+
+envoy_stream_t Dispatcher::startStream(envoy_observer observer) {
+  envoy_stream_t new_stream_id = current_stream_id_++; // FIXME: This needs to be equivalent to
+                                                       // getAnIncrement(); confirm or update.
+
+  event_dispatcher_.post([this, observer, new_stream_id]() -> void {
+    DirectStreamCallbacksPtr callbacks =
+        std::make_unique<DirectStreamCallbacks>(new_stream_id, observer, *this);
+    AsyncClient& async_client = cluster_manager_.httpAsyncClientForCluster("egress_cluster");
+    AsyncClient::Stream* underlying_stream = async_client.start(*callbacks, {});
+
+    DirectStreamPtr direct_stream =
+        std::make_unique<DirectStream>(underlying_stream, std::move(callbacks));
```

Comment:
Assuming we plan on handling null above, prefer passing a reference to underlying_stream here vs. pointer.